[Provisioner] Generate valid cluster names when username has invalid characters

romilbhardwaj opened the pull request:
Closes #1161.

We now clean up the username (used as suffix in the cluster name) by:

1. Making all characters lowercase
2. Removing any non-alphanumeric characters (excluding hyphens)
3. Removing any numbers and/or hyphens at the start of the username.
4. Removing any hyphens at the end of the username

e.g. `1SkY-PiLot2-` becomes `sky-pilot2` when it is used for generating cluster name.

## Tested:

Manual tests by overriding the username read by getpass.getuser():
- [x] `LOGNAME=RomilB sky launch` - uppercase characters
- [x] `LOGNAME=romil-b sky launch` - hypen
- [x] `LOGNAME=romil_b sky launch` - underscore
- [x] `LOGNAME=1roMil-b123 sky launch` - number at start gets removed. Becomes `romilb123`
- [x] `LOGNAME=r#omIl_B sky launch`

concretevitamin reviewed (commented):
Thanks @romilbhardwaj!

1. How about keeping hyphens in a username? Since they are used and allowed by clouds.
2. One other place where non-conforming username may be picked up as part of a cluster name is `_default_interactive_node_name()` in cli.py. Should we fix that too?
3. There are a few other usage of `getpass.getuser()` but I think only the above two are for cluster names purposes.

romilbhardwaj commented:
These are great points:

> 1. How about keeping hyphens in a username? Since they are used and allowed by clouds.

I have updated the code to allow hyphens in the middle the of the username, but not at start or end. Thinking more about it, maybe it's better to have a simpler scheme (disallow hyphens everywhere), but I don't feel strongly either way.

> 2. One other place where non-conforming username may be picked up as part of a cluster name is `_default_interactive_node_name()` in cli.py. Should we fix that too?

Great catch! I have updated `_default_interactive_node_name()`

> 3. There are a few other usage of `getpass.getuser()` but I think only the above two are for cluster names purposes.

Yeah, there are some used in hashing job ids, but perhaps they don't need to be changed. Changing them would require some thought about backward compatibility.

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1127,7 +1127,27 @@ def check_local_gpus() -> bool:
+     3. Removing any numbers and/or hyphens at the start of the username.
+     4. Removing any hyphens at the end of the username
+
+    e.g. 1SkY-PiLot2- becomes sky-pilot2.
```
```suggestion
    e.g. 1SkY-PiLot2- becomes sky-pilot2.
    
    Returns:
      A cleaned username that will pass the regex in check_cluster_name_is_valid().
```

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -1127,7 +1127,27 @@ def check_local_gpus() -> bool:
+
+
+def get_cleaned_username() -> str:
+    """
+    Cleans the current username to be used as a cluster name.
```
```suggestion
    """Cleans the current username to be used as part of a cluster name.
```
Styleguide

concretevitamin reviewed (approved):
Thanks @romilbhardwaj. LGTM.
